### PR TITLE
Issue 20369 bash script imports

### LINKED ISSUE #20369: We need to create a bash script that imports Content types in our native json format
The customer should be able to pulldown a dotCMS git repo with a folder containing the definition of CT and import them from a dotCMS instance via the CLT Which internally relies on the _**/api/v1/contenttype**_  endpoint to do the import/export 

The CLT will use internally CURL But it will initially be written in bash. or python. 
The idea is to be able to checkout a directory from say github and then push that directory into a dotCMS instance.

Define a folder structure that we can also use when generating bundles that can be used to push these assets into a dotCMS instance.

Best efforts for knowable IDs:  We will need to update the creation of sites, content types, folders and files and use a sha256 of their hostName+language+path as the identifier for these assets.  If that hash already exists in the db, we will just create a new UUID and use it instead.
See: #19443

Pass a directory  and push its content into a dotCMS instance.
1. Sites 
   - file extension pattern = `.site.json`
   - add site based on siteName
   - include identifier in the json
   
2. ContentTypes (should be the same format as we get from the API)
   - file extension pattern = `.contenttype.json`
   - do a find on the passed in directory for
   - if content type exists (by var) then we update
   - else we put a new content type
   
3. File Assets - via webdav (can reuse the webdav endpoint)
   - https://dotcms.com/docs/latest/scripting-webdav
   - Best efforts for knowable IDs:  We will need to update the creation of folders and files and use a sha256 of their hostName+language+path as the identifier for these assets.  If that hash already exists in the db, we will just create a new UUID and use it instead.
   - Use the bundle layout found here: https://docs.google.com/spreadsheets/d/1iO7lq23Rdcev_-sFB5WdX8-KrEsy96dUTdMhvHY8IrE/edit#gid=0

The scripts/files should accept ContentTypes/fields with and without inodes.  If the contenttype definition includes the inode  and they do not match the ones we have in the dotCMS instance, the script should fail.

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
This PR gathers changes meant to generate a predictable unique identifier + a CLI tool to simplify pushing content via the command line.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierAPIImpl.java [hunk 1]
```diff
@@ -141,24 +166,105 @@ public Identifier createNew(final Versionable asset, final Treeable parent) thro
...
+	
+	@VisibleForTesting
+    @CloseDBIfOpened
+    String bestEffortsKnowableIdHash(final String stringToHash) {
+        
+        final String hashed = DigestUtils.sha256Hex(stringToHash.toLowerCase()).substring(0,32);
+        Logger.info(this.getClass(), "");
+        Logger.info(this.getClass(), "hashing id: " + stringToHash + " to " + hashed);
+        Logger.info(this.getClass(), "");
+        if(new DotConnect()
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierAPIImpl.java:239]: very nice

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/business/IdentifierAPIImpl.java:239]: and yeah, this should be a separated class

**Review by jdotcms — COMMENTED** [review]
CLi commands looks very good, just remember we have to push after same 
both: sites and ct

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java [hunk 2]
```diff
@@ -479,51 +467,48 @@ protected Identifier find(final String identifier) throws DotStateException, Dot
...
-			try{
-				dc.loadResult();
-			}catch(DotDataException e){
-				Logger.error(IdentifierFactoryImpl.class, "saveIdentifier failed:" + e, e);
-				throw new DotDataException(e);
-			}
+        String query;
+        if (UtilMethods.isSet(id.getId())) {
+            if (isIdentifier(id.getId())) {
+                query = "UPDATE identifier set parent_path=?, asset_name=?, host_inode=?, asset_type=?, syspublish_date=?, sysexpire_date=?, owner=?, create_date=?, asset_subtype=? where id=?";
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:479]: these scripts (insert/update) can be declared as private constants

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java [hunk 3]
```diff
@@ -135,6 +138,71 @@ public final Response saveLanguage(@Context final HttpServletRequest request,
         return Response.ok(new ResponseEntityView(language)).build(); // 200
     }
 
+    @POST
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:141]: I think some curl tests can be included

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java [hunk 4]
```diff
@@ -135,6 +138,71 @@ public final Response saveLanguage(@Context final HttpServletRequest request,
...
+        this.webResource.init(null, request, response,
+                true, PortletID.LANGUAGES.toString());
+
+        final Locale locale = validateLanguageTag(languageTag);
+
+        final LanguageForm languageForm = new LanguageForm.Builder()
+                .language(locale.getDisplayLanguage()).languageCode(locale.getLanguage())
+                .country(locale.getDisplayCountry()).countryCode(locale.getCountry()).build();
+
+        final Language language = saveOrUpdateLanguage(null, languageForm);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:160]: the saveOrUpdateLanguage, should be able to validate if do not exists any language with that combination
see here for more reference: https://github.com/dotcms/core/issues/7342

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java [hunk 5]
```diff
@@ -135,6 +138,71 @@ public final Response saveLanguage(@Context final HttpServletRequest request,
...
+    @Path("/{languageTag}")
+    @JSONP
+    @NoCache
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public Response getFromLanguageTag (
+            @Context final HttpServletRequest request,
+            @Context final HttpServletResponse response,
+            @PathParam("languageTag") final String languageTag) {
+
+        final InitDataObject initData = new WebResource.InitBuilder(webResource)
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:175]: why doing doble init?

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java [hunk 6]
```diff
@@ -135,6 +138,71 @@ public final Response saveLanguage(@Context final HttpServletRequest request,
...
+
+    }
+
+    private Locale validateLanguageTag(final String languageTag)throws DoesNotExistException {
+        final Locale locale = Locale.forLanguageTag(languageTag);
+        final boolean validCountry = (isNotSet(locale.getCountry()) || Stream.of(Locale.getISOCountries()).collect(Collectors.toSet()).contains(locale.getCountry()));
+        final boolean validLang = Stream.of(Locale.getISOLanguages()).collect(Collectors.toSet()).contains(locale.getLanguage());
+        if(validLang && validCountry) {
+            return locale;
+        }else throw new DoesNotExistException(String.format(" `%s` is an invalid language tag ", languageTag));
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:203]: use curly brackets

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPI.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package com.dotmarketing.business;
+
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotmarketing.portlets.languagesmanager.model.Language;
+import java.util.function.Supplier;
+
+public interface DeterministicIdentifierAPI {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPI.java:7]: I would add some doc explaining the nature of this API

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+        Logger.warn(DeterministicIdentifierAPIImpl.class,String.format("Attempt to generate content-type id has failed instead a non deterministic id (%s) was returned.",candidateId));
+        return candidateId;
+    }
+
+    /**
+     * Test if the calculated hash has already been used
+     * @param hash
+     * @return
+     */
+    private boolean isContentTypeInode(final String hash){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:260]: I guess it needs CloseDBIfOpened as others

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+
+    /**
+     * Given a Language this will evaluate the code and country code if any then generate a sha256 and finally will hash it out into a long val
+     * @param lang
+     * @return
+     */
+    @Override
+    public long generateDeterministicIdBestEffort(final Language lang) {
+        //You never know.. better to be safe than sorry
+        if("US".equals(lang.getCountryCode()) &&  "en".equals(lang.getLanguageCode())){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:305]: @wezell not sure if this is all times true, we know some clients deletes all langs and start from scratch, so may be should be a config prop to turn on, turn off

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+    private synchronized long nextLangId(){
+        return System.currentTimeMillis();
+    }
+
+    /**
+     * Test if the calculated hash has already been used
+     * @param hash
+     * @return
+     */
+    private boolean isLanguageId(final long hash){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:335]: add CloseDB

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+                .getInt("test")>0;
+    }
+
+    /**
+     * basically takes a string and hash it into a long number
+     * Used to create a deterministic long representation out of a sha256
+     * @param string
+     * @return
+     */
+    private long longHash(final String string) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:348]: nice

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:348]: not sure if this should be expose in other place, may be UUIDUtils

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java (hunk 9 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:305]: FYI:  clients could start dotcms with a diff default language than `en_US`

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:305]: Removing it

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java (hunk 3 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:141]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java (hunk 4 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:160]: done

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPI.java (hunk 7 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPI.java:7]: done

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java (hunk 11 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:348]: I think it's fine here

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java (hunk 8 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:260]: done. However, these methods are consumed from another which is already marked with CloseDBIfOpened

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java (hunk 6 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:199]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java (hunk 5 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java]: fixed. it was a mistake

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java (hunk 2 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:479]: they could. but this class is All formatted following that pattern I'd rather leave it like so

File: dotCMS/src/curl-test/LanguageResourceTests.json [hunk 12]
```diff
@@ -436,7 +683,86 @@
 						{
 							"listen": "prerequest",
 							"script": {
-								"id": "5cc4ea17-d93c-4fc0-9f60-26b33dce679b",
+								"exec": [
```
Comment (nollymar) [inline dotCMS/src/curl-test/LanguageResourceTests.json:686]: Please add a description to this test

File: dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeFactoryImpl.java [hunk 13]
```diff
@@ -686,25 +718,42 @@ private void deleteContentletsByType(ContentType type) throws DotDataException {
...
+                .byContentTypeIdAndVar(rel.getParentStructureInode(), rel.getChildRelationName());
+        contentTypeFieldAPI.delete(fieldToDelete);
       }
-      APILocator.getRelationshipAPI().delete(rel);
+      relationshipAPI.delete(rel);
+    }
+
+    //Once the parent - child : child - parent relationship has been cleared
+    //Now we need to remove the field from the actual CT.
+    final List<Field> fields = type.fields(RelationshipField.class);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeFactoryImpl.java:748]: I am not sure this code fixes the root cause of the problem. I think it fixes the broken test, but not the problem in the existing logic

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeFactoryImpl.java:748]: I'll look into it once again. But I think it solves the problem of removing hanging relationship fields. The problem revealed by the test. The code isn't only meant to fix a test.

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeFactoryImpl.java]: The logic was faulty got it fixed Per @nollymar hunch

**Review by nollymar — APPROVED** [review]

**Pull request merged**